src/lib.rs

```diff
@@ -39,8 +39,48 @@
 #![crate_type = "rlib"]
 #![feature(core_intrinsics)]
 
+use std::{
+    alloc::{Layout, GlobalAlloc},
+    ops::Deref,
+    sync::atomic::{AtomicBool, Ordering},
+};
 
 static SIZE_ALLOC_INFO: usize = (1024 * 1024) * 2; // 2KB
+/// A spinlock for the global ALLOC_INFO list.
+///
+/// System allocators are generally thread-safe, but gcmalloc reads and writes
+/// metadata about each allocation to shared memory upon returning from the
+/// system allocator call. This requires additional synchronisation mechanisms.
+/// It is not possible to use `sys::Sync::Mutex` for this purpose as the
+/// implementation includes a heap allocation containing a `pthread_mutex`. This
+/// would cause infinite recursion when initialising the lock as the allocator
+/// would call itself.
+///
+/// Instead, we use a simple global spinlock built on an atomic bool to guard
+/// access to the ALLOC_INFO list.
+static ALLOC_LOCK: AllocLock = AllocLock(AtomicBool::new(false));
+
+struct AllocLock(AtomicBool);
+
+impl AllocLock {
+    fn lock(&self) {
+        while !ALLOC_LOCK.compare_and_swap(false, true, Ordering::AcqRel) {
```

Comment:
I think you may have mentioned this, but shouldn't it be:
```
while !self.0.compare_and_swap(...
```

i.e. `self`.

Reply:
Yes, I was testing you